global-setup.e2e.ts

```diff
@@ -1,44 +1,68 @@
-import { execFileSync, execSync } from 'child_process';
+import { join } from 'node:path';
 import { setup as globalSetup } from './global-setup';
 import { setupTestFolder, teardownTestFolder } from './testing/test-setup/src';
-import startLocalRegistry from './tools/scripts/start-local-registry';
-import stopLocalRegistry from './tools/scripts/stop-local-registry';
+import {
+  nxRunManyNpmInstall,
+  nxRunManyNpmUninstall,
+} from './tools/src/npm/utils';
+import { findLatestVersion, nxRunManyPublish } from './tools/src/publish/utils';
+import startLocalRegistry from './tools/src/verdaccio/start-local-registry';
+import stopLocalRegistry from './tools/src/verdaccio/stop-local-registry';
+import { RegistryResult } from './tools/src/verdaccio/types';
 
-const localRegistryNxTarget = '@code-pushup/cli-source:local-registry';
+const uniquePort: number = Number(
+  (6000 + Number(Math.random() * 1000)).toFixed(0),
+);
+const e2eDir = join('tmp', 'e2e');
+const uniqueDir = join(e2eDir, `registry-${uniquePort}`);
+
+let activeRegistry: RegistryResult;
 
 export async function setup() {
   await globalSetup();
   await setupTestFolder('tmp/local-registry');
-  await setupTestFolder('tmp/e2e');
+  await setupTestFolder(e2eDir);
 
   try {
-    await startLocalRegistry({ localRegistryTarget: localRegistryNxTarget });
+    activeRegistry = await startLocalRegistry({
+      localRegistryTarget: '@code-pushup/cli-source:start-verdaccio',
+      storage: join(uniqueDir, 'storage'),
+      port: uniquePort,
+    });
   } catch (error) {
     console.error('Error starting local verdaccio registry:\n' + error.message);
     throw error;
   }
 
+  // package publish
+  const { registry } = activeRegistry.registryData;
+  try {
+    console.info('Publish packages');
+    nxRunManyPublish({ registry, nextVersion: findLatestVersion() });
+  } catch (error) {
+    console.error('Error publishing packages:\n' + error.message);
+    throw error;
+  }
+
+  // package install
   try {
     console.info('Installing packages');
-    execFileSync(
-      'npx',
-      ['nx', 'run-many', '--targets=npm-install', '--parallel=1'],
-      { env: process.env, stdio: 'inherit', shell: true },
-    );
+    nxRunManyNpmInstall({ registry });
   } catch (error) {
     console.error('Error installing packages:\n' + error.message);
     throw error;
   }
 }
 
 export async function teardown() {
-  stopLocalRegistry();
-  console.info('Uninstalling packages');
-  execFileSync(
-    'npx',
-    ['nx', 'run-many', '--targets=npm-uninstall', '--parallel=1'],
-    { env: process.env, stdio: 'inherit', shell: true },
-  );
-  await teardownTestFolder('tmp/e2e');
-  await teardownTestFolder('tmp/local-registry');
+  if (activeRegistry && 'registryData' in activeRegistry) {
+    const { stop } = activeRegistry;
+
+    stopLocalRegistry(stop);
+    nxRunManyNpmUninstall();
+  } else {
+    activeRegistry.stop();
+    return;
+  }
```

Comment:
This `else`-block seems ill-conceived. If `activeRegistry` isn't set or doesn't have a `registryData` property, then I would expect `activeRegistry.stop()` to throw a runtime error - hence the type error. And I don't see why should skip cleaning up the temporary test folder in such a case.

![image](https://github.com/user-attachments/assets/9b12390e-e0de-49a7-bc32-7ec8e6707810)

Should probably be removed:

```suggestion
  }
```

Or if you want to log it:

```suggestion
  } else {
    console.warn('Active registry not found during teardown.');
  }
```